billing-report-agent/src/main/java/com/epam/pipeline/billingreportagent/service/impl/converter/RunToBillingRequestConverter.java

```diff
@@ -0,0 +1,172 @@
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Data
+@Slf4j
+public class RunToBillingRequestConverter implements EntityToBillingRequestConverter<PipelineRun> {
+
+    private final String indexPrefix;
+    private final EntityMapper<PipelineRunBillingInfo> mapper;
+
+    public RunToBillingRequestConverter(final String indexPrefix,
+                                        final EntityMapper<PipelineRunBillingInfo> mapper) {
+        this.indexPrefix = indexPrefix;
+        this.mapper = mapper;
+    }
+    /**
+     * Creates billing requests for given run
+     * @param container Pipeline run to build request
+     * @param indexName index to insert requests into
+     * @param syncStart time point, where the whole synchronization process was started
+     * @return list of requests to be performed (deletion index request if no billing requests created)
+     */
+    @Override
+    public List<DocWriteRequest> convertEntityToRequests(final EntityContainer<PipelineRun> container,
+                                                         final String indexName,
+                                                         final LocalDateTime syncStart) {
+            return convertRunToBillings(container, syncStart).stream()
+                .map(billingInfo -> getDocWriteRequest(indexName, container, billingInfo))
+                .collect(Collectors.toList());
+    }
+
+    protected List<PipelineRunBillingInfo> convertRunToBillings(final EntityContainer<PipelineRun> run,
+                                                                final LocalDateTime syncStart) {
+        final BigDecimal pricePerHour = run.getEntity().getPricePerHour();
+        final List<RunStatus> statuses = run.getEntity().getRunStatuses();
+        adjustStatuses(statuses, syncStart, run.getEntity().getId());
+        final List<List<RunStatus>> activePeriods = calculateActivityPeriods(statuses);
+
+        final Map<LocalDate, PipelineRunBillingInfo> reports = new HashMap<>();
+        activePeriods.forEach(period -> {
+            for (int i = 0; i < period.size() - 1; i++) {
+                final LocalDateTime start = period.get(i).getTimestamp();
+                final LocalDateTime end = period.get(i + 1).getTimestamp();
+                final Long dailyCost = calculateCostsForPeriod(start, end, pricePerHour);
+                addDailyCostReport(reports, start.toLocalDate(), dailyCost, ChronoUnit.MINUTES.between(start, end), run);
+            }
+        });
+        return new ArrayList<>(reports.values());
+    }
+
+    private void adjustStatuses(final List<RunStatus> statuses, final LocalDateTime syncStart, final Long runId) {
+        statuses.sort(Comparator.comparing(RunStatus::getTimestamp));
```

Comment:
If statuses are not present let's assume run was active from `startDate` till `endDate`

Reply:
Done